File: Website/js/Calculator.js
//Calculate is the function that checks the validators.js file to validate the data
//and then handles the math.
function calculate(e) {
    e.preventDefault();

    // ***** VARIABLES *****

    // The following 3 statements instantiate validated form inputs to their variables.
    let homeCost = Number.parseFloat(document.getElementById("homeCost").value);
    if (!myHomeCostValidationFunction()) {
        console.log("invalid");
        return
    }
    let downPayment = Number.parseFloat(document.getElementById("downPayment").value);
    if (!myDownPaymentValidationFunction()) {
        console.log("invalid");
        return
    }
    let apr = Number.parseFloat(document.getElementById("apr").value);
    if (!myAprValidationFunction()) {
        console.log("invalid");
        return
    }
    // years is taken in from the drop down menu.
    let years = Number.parseInt(document.getElementById("years").value);
    // monthlyInterest is instantiated as a decimal.
    let monthlyInterest = apr / 100 / 12;
    // Calculates total loan payments.
    let monthlyPayments = years * 12;
    // Sets the value of the loan.
    let principal = homeCost - downPayment;
    // remainingPrincipal has to be set to principal at the start of the loop and 
    // it is decremented by the principal paid.
    let remainingPrincipal = principal;
    // Determines whether the box that will run the loop for the amortization 
    // schedule is checked.  
    let amortSched = document.getElementById("amortSched").checked;




    // ***** MATH *****

    // Calculate monthly payment using principal, monthly apr and loan duration in months.
    mortgagePayment = principal * (monthlyInterest * Math.pow(1 + monthlyInterest, monthlyPayments)
        / (Math.pow(1 + monthlyInterest, monthlyPayments) - 1));
    document.getElementById("principal").innerHTML = currencyFormat(principal);
    document.getElementById("mortgagePayment").innerHTML = currencyFormat(mortgagePayment);

    // Calculates total cost of the loan.
    totalPayments = mortgagePayment * monthlyPayments;
    document.getElementById("totalPayments").innerHTML = currencyFormat(totalPayments);

    // Calculates total interest paid.
    totalInterest = totalPayments - principal;
    document.getElementById("totalInterest").innerHTML = currencyFormat(totalInterest);

    // Instantiates the output object outside the loop.
    document.getElementById("output").innerHTML = "";




    // ***** LOOP *****

    // If the check box is selected, the loop runs.  Each iteration produces for numbers:
    // payment number, remaining principal, the portion of the payment that is interest
    // and the portion of the payment that is principal.  The principal should increase
    // top to bottom and the interest should decrease top to bottom.
    if (amortSched == true) {

        // The best work around I could find to get both the data validation and the output
        // number formatting to work was to print out to the HTML from the JS file.  As a 
        // result I needed to insert the header here as well, just above the loop.
        document.getElementById("output").innerHTML += `<tr>
        <td>Payment</td>
        <td>Remaining Principal</td>
        <td>Monthly Interest</td>
        <td>Monthly Principal</td>
        </tr>`

        // Getting payments from 1 to number of payments requires adding another month 
        // because the numbers are indexed from zero.
        for (i = 1; i < (monthlyPayments + 1); i++) {

            // Calculates monthly interest cost
            monthlyAmortInt = remainingPrincipal * monthlyInterest;

            // Reduces var remainingPrincipal to the principal left over after the interest
            // is paid.
            remainingPrincipal = remainingPrincipal - (mortgagePayment - monthlyAmortInt);

            // Calculates monthly principal paid.
            monthlyPrincipal = mortgagePayment - monthlyAmortInt;

            // Prints month, principal remaining, monthly interest and monthly principal
            // on one line as my output object
            document.getElementById("output").innerHTML += `<tr>
            <td>${i}</td>
            <td>${currencyFormat(remainingPrincipal)}</td>
            <td>${currencyFormat(monthlyAmortInt)}</td>
            <td>${currencyFormat(monthlyPrincipal)}</td>
          </tr>`
        }
    }
}


// This is the listener for the submit button that runs the calculate function
document.querySelector("form").addEventListener("submit", calculate);

File: Java Command Line Calculator/calculator.java
package mortgageCalculator;

import java.text.NumberFormat;
import java.util.Scanner;

public class calculator {

	public static void main(String[] args) {
		
		// Initialize new scanner keyboard input
		Scanner scanner = new Scanner(System.in);

		//Because they were in the try block the variables are initiated here
		double homeCost = 0.0;
		double downPayment = 0.0;
		double apr = 0.0;
		double monthlyInterest = 0.0;
		int years = 0;
		int monthlyPayments = 0;	
		boolean askForInput = true;

		// While loop to restart the process in the event of an invalid input
		while (askForInput)
			try {
				// Input cost
				System.out.print("Home Cost:  $");
				homeCost = scanner.nextDouble();

				// Input down payment
				System.out.print("Down Payment:  $");
				downPayment = scanner.nextDouble();

				// Input annual percentage rate
				System.out.print("APR:  %");
				apr = scanner.nextDouble();
				// /100 to get to decimal and /12 to get to monthly val
				monthlyInterest = apr / 100 / 12;

				// Input duration and convert to months for calculation
				System.out.print("Duration (whole years):");
				years = scanner.nextInt();
				monthlyPayments = years * 12;
				askForInput = false;
			}

			// Not clearing the scanner blew up the process so the clear buffer is the work around
			catch (Exception e) {
				String clearBuffer = scanner.next();
				System.out.println("Invalid Input - Please Try Again");
			}

		// Calculate loan value
		double principal = homeCost - downPayment;

		// Calculate monthly payment and converts to currency
		double mortgagePayment = principal * (monthlyInterest * Math.pow(1 + monthlyInterest, monthlyPayments)
				/ (Math.pow(1 + monthlyInterest, monthlyPayments) - 1));
		String mortgageFormat = NumberFormat.getCurrencyInstance().format(mortgagePayment);

		// Calculates total cost of the loan and converts to money
		double totalPayments = mortgagePayment * monthlyPayments;
		String paymentFormat = NumberFormat.getCurrencyInstance().format(totalPayments);

		// Displays monthly payment and loan cost
		System.out.println("Monthly Payment:  " + mortgageFormat);
		System.out.println("Total Loan Cost:  " + paymentFormat);

		// Initialize amortTable with no value
		String amortTable = "";
		
		// While loop exists to start over in the event of an invalid input
		while (!amortTable.toUpperCase().equals("Y")) {
			// User inputs Y or N to see the amortization table
			System.out.print("Would you like to see the amortization schedule (Y or N)?");
			amortTable = scanner.next();

			// Initializes the variable remaining principal to the value of the loan
			double remainingPrincipal;
			remainingPrincipal = principal;

			// Loop to display the amortization data by month
			if (amortTable.toUpperCase().equals("Y")) {
				System.out.println("\nMonth    Remaining Principal:  "
						+ "    Monthly Interest:  "
						+ "     monthly Principal:  " );
				for (int i = 1; i < (monthlyPayments + 1); i++) {
				
					// Calculates monthly interest cost
					double monthlyAmortInt = remainingPrincipal * monthlyInterest;
					
					// Reduces var remainingPrincipal to the principal left over after the interest
					// is paid
					remainingPrincipal = remainingPrincipal - (mortgagePayment - monthlyAmortInt);
					
					// Calculates monthly principal paid
					double monthlyPrincipal = mortgagePayment - monthlyAmortInt;
					
					// Prints month, principal remaining, monthly interest and monthly principal
					System.out.printf( "%5d         $%10.2f            $%8.2f             $%8.2f\n ", i, 
							remainingPrincipal, monthlyAmortInt, monthlyPrincipal);
				}
			} else if (amortTable.toUpperCase().equals("N"))
			{
				// Exits if user doesn't want to see table
				System.out.println("Thank you");
				System.exit(0);
			}
			else {
				System.out.println("Invalid Input");
			}
		}
		scanner.close();
	}
}
